### ISSUE TITLE
Preprocess tactic gets "stuck" visiting letrecs sometimes

### ISSUE THREAD

**Issue by amosr** [FStarLang/FStar#3210]
Hi again,

I'm running into another error with preprocess_tac, this time with letrecs. I think it's a separate issue from #3209, which is related to records.

I don't think this is specific to visit, but I haven't been able to minimise it further yet.

```
module Test.VisitLetRec

module Tac = FStar.Tactics

#push-options "--trace_error --error_contexts true"

let lift_ty: Tac.term -> Tac.Tac Tac.term =
  let rec go (t: Tac.term): Tac.Tac Tac.term =
    t
  in Tac.visit_tm go

[@@Tac.preprocess_with lift_ty]
let rec count (x: int): int =
  if x > 0 then count (x - 1) else 0

[@@Tac.preprocess_with lift_ty]
let test_localletrec (x: int): int =
  let rec count x = if x > 0 then count (x - 1) else 0 in
  count 0
```

I get the error:
```
Unexpected error
FStar_Errors.Error(_)
Raised at FStar_Compiler_Effect.raise in file "fstar-lib/FStar_Compiler_Effect.ml" (inlined), line 10, characters 12-17
Called from FStar_Syntax_Unionfind.with_uf_enabled in file "fstar-lib/generated/FStar_Syntax_Unionfind.ml", line 110, characters 40-74
Called from FStar_TypeChecker_Tc.tc_decls in file "fstar-lib/generated/FStar_TypeChecker_Tc.ml", line 4907, characters 8-170
Called from FStar_TypeChecker_Tc.tc_partial_modul.(fun) in file "fstar-lib/generated/FStar_TypeChecker_Tc.ml", line 5095, characters 25-77
Called from FStar_TypeChecker_Tc.tc_modul in file "fstar-lib/generated/FStar_TypeChecker_Tc.ml", line 5175, characters 20-44
Called from FStar_TypeChecker_Tc.check_module in file "fstar-lib/generated/FStar_TypeChecker_Tc.ml", line 5355, characters 22-39
Called from FStar_Universal.tc_one_file.tc_source_file.check_mod.check.(fun) in file "fstar-lib/generated/FStar_Universal.ml", line 1086, characters 29-141
Called from FStar_Universal.with_tcenv_of_env in file "fstar-lib/generated/FStar_Universal.ml", line 130, characters 65-73
Called from FStar_Universal.tc_one_file.tc_source_file.check_mod in file "fstar-lib/generated/FStar_Universal.ml", line 1107, characters 21-134
Called from FStar_Universal.tc_one_file in file "fstar-lib/generated/FStar_Universal.ml", line 1180, characters 32-49
Called from FStar_Universal.tc_one_file_from_remaining in file "fstar-lib/generated/FStar_Universal.ml", line 1328, characters 16-98
Called from FStar_Universal.tc_fold_interleave in file "fstar-lib/generated/FStar_Universal.ml", line 1360, characters 19-71
...


* Warning 266 at FStar.Tactics.Effect.fsti(89,11-100,32):
  - Unembedding failed for type
    FStar.Stubs.Tactics.Result.__result FStar.Stubs.Reflection.Types.term
  - emb_typ =
    (FStar.Stubs.Tactics.Result.__result FStar.Stubs.Reflection.Types.term)
  - Term =
    FStar.Stubs.Tactics.Result.Success
      (FStar.Stubs.Reflection.V2.Builtins.pack_ln (FStar.Stubs.Reflection.V2.Data.Tv_Abs
              (FStar.Stubs.Reflection.V2.Builtins.pack_binder (FStar.Stubs.Reflection.V2.Data.Mkbinder_view
                      (`Prims.int)
                      FStar.Stubs.Reflection.V2.Data.Q_Explicit
                      []
                      (FStar.Sealed.seal "x")))
              (FStar.Stubs.Reflection.V2.Builtins.pack_ln (FStar.Stubs.Reflection.V2.Data.Tv_AscribedC
                      (FStar.Stubs.Reflection.V2.Builtins.pack_ln (FStar.Stubs.Reflection.V2.Data.Tv_Match
                              (`Prims.op_GreaterThan x 0 <: Prims.bool)
                              FStar.Pervasives.Native.None
                              [
                                FStar.Stubs.Reflection.V2.Data.Pat_Constant
                                FStar.Stubs.Reflection.V2.Data.C_True,
                                (`Test.VisitLetRec.count (Prims.op_Subtraction x
                                        1));
                                FStar.Stubs.Reflection.V2.Data.Pat_Var
                                  (FStar.Sealed.seal (`_)) "uu___",
                                (`0)
                              ]))
                      (FStar.Stubs.Reflection.V2.Builtins.pack_comp (FStar.Stubs.Reflection.V2.Data.C_Total
                            (`Prims.int)))
                      FStar.Pervasives.Native.None
                      false)))) "(((proofstate)))"
  - Stack trace:
    Raised by primitive operation at FStar_Compiler_Util.stack_dump in file "fstar-lib/FStar_Compiler_Util.ml", line 129, characters 53-82
    Called from FStar_Errors_Msg.backtrace_doc in file "fstar-lib/generated/FStar_Errors_Msg.ml", line 43, characters 12-45
    Called from FStar_Errors.maybe_add_backtrace in file "fstar-lib/generated/FStar_Errors.ml", line 824, characters 32-65
    Called from FStar_Errors.log_issue_ctx in file "fstar-lib/generated/FStar_Errors.ml", line 837, characters 23-46
    Called from FStar_Syntax_Embeddings_Base.unembed in file "fstar-lib/generated/FStar_Syntax_Embeddings_Base.ml", line 243, characters 11-70
    Called from FStar_Tactics_Interpreter.unembed in file "fstar-lib/generated/FStar_Tactics_Interpreter.ml" (inlined), line 27, characters 21-74
    Called from FStar_Tactics_Interpreter.unembed_tactic_0.(fun) in file "fstar-lib/generated/FStar_Tactics_Interpreter.ml", line 157, characters 28-114
...

* Error 170 at shelve/bug-reports/Test.VisitLetRec.fst(13,0-14,36):
  - Tactic failed
  - Tactic got stuck!
  - Reduction stopped at: FStar.Stubs.Tactics.Result.Success
      (FStar.Stubs.Reflection.V2.Builtins.pack_ln (FStar.Stubs.Reflection.V2.Data.Tv_Abs
              (FStar.Stubs.Reflection.V2.Builtins.pack_binder (FStar.Stubs.Reflection.V2.Data.Mkbinder_view
                      (`Prims.int)
                      FStar.Stubs.Reflection.V2.Data.Q_Explicit
                      []
                      (FStar.Sealed.seal "x")))
              (FStar.Stubs.Reflection.V2.Builtins.pack_ln (FStar.Stubs.Reflection.V2.Data.Tv_AscribedC
                      (FStar.Stubs.Reflection.V2.Builtins.pack_ln (FStar.Stubs.Reflection.V2.Data.Tv_Match
                              (`Prims.op_GreaterThan x 0 <: Prims.bool)
                              FStar.Pervasives.Native.None
                              [
                                FStar.Stubs.Reflection.V2.Data.Pat_Constant
                                FStar.Stubs.Reflection.V2.Data.C_True,
                                (`Test.VisitLetRec.count (Prims.op_Subtraction x
                                        1));
                                FStar.Stubs.Reflection.V2.Data.Pat_Var
                                  (FStar.Sealed.seal (`_)) "uu___",
                                (`0)
                              ]))
                      (FStar.Stubs.Reflection.V2.Builtins.pack_comp (FStar.Stubs.Reflection.V2.Data.C_Total
                            (`Prims.int)))
                      FStar.Pervasives.Native.None
                      false)))) "(((proofstate)))"
  - Stack trace:
    Raised by primitive operation at FStar_Compiler_Util.stack_dump in file "fstar-lib/FStar_Compiler_Util.ml", line 129, characters 53-82
    Called from FStar_Errors_Msg.backtrace_doc in file "fstar-lib/generated/FStar_Errors_Msg.ml", line 43, characters 12-45
    Called from FStar_Errors.maybe_add_backtrace in file "fstar-lib/generated/FStar_Errors.ml", line 824, characters 32-65
    Called from FStar_Errors.raise_error_doc in file "fstar-lib/generated/FStar_Errors.ml", line 939, characters 27-50
    Called from FStar_Tactics_Monad.bind.(fun) in file "fstar-lib/generated/FStar_Tactics_Monad.ml", line 265, characters 28-33
    Called from FStar_Tactics_Monad.run in file "fstar-lib/generated/FStar_Tactics_Monad.ml" (inlined), line 241, characters 23-33
    Called from FStar_Tactics_Monad.run_safe.(fun) in file "fstar-lib/generated/FStar_Tactics_Monad.ml", line 253, characters 50-58
    Called from FStar_Tactics_Interpreter.run_tactic_on_ps' in file "fstar-lib/generated/FStar_Tactics_Interpreter.ml", line 925, characters 24-281
    Called from FStar_Tactics_Hooks.mpreprocess.(fun) in file "fstar-lib/generated/FStar_Tactics_Hooks.ml", line 2309, characters 19-313
    Called from FStar_Errors.with_ctx in file "fstar-lib/generated/FStar_Errors.ml", line 1013, characters 27-31
    Called from FStar_TypeChecker_Env.preprocess in file "fstar-lib/generated/FStar_TypeChecker_Env.ml" (inlined), line 1483, characters 37-65
...

```

I will update this issue as I continue to debug.

**Comment by amosr** [comment]
I've minimised it to a smaller test case.

```
module Test.VisitLetRec

module Ref = FStar.Reflection.V2
module Tac = FStar.Tactics.V2
open FStar.Tactics.Effect
open FStar.Tactics.Util

module Visit = FStar.Tactics.Visit

let rec repack_tm (t : Ref.term) : Tac Ref.term =
  let tv = Ref.inspect_ln t in
  let tv' =
    match tv with
    | Ref.Tv_Match sc ret_opt brs ->
        let brs = map repack_br brs in
        Tac.print (Tac.term_to_string (quote brs));
        Ref.Tv_Match sc ret_opt brs
    | _ -> tv
  in
  Ref.pack_ln tv'
and repack_br (b:Ref.branch) : Tac Ref.branch =
  let (p, t) = b in
  let t = repack_tm t in
  match p with
  | Ref.Pat_Var s v ->
    // let vv = Ref.as_ppname "uu___" in
    (Ref.Pat_Var (FStar.Sealed.seal (`_)) v, t)
  | _ -> (p, t)

[@@preprocess_with repack_tm]
let count: unit =
  match () with | _ -> ()
```

This prints  the following:
```
TAC>> [FStar.Stubs.Reflection.V2.Data.Pat_Var (FStar.Sealed.seal (`_)) "uu___", (`())]
```
Note that the "uu___" is unsealed. If we construct a new ppname with `as_ppname`, the result is sealed

**Comment by mtzguido** [comment]
Fixed by #3211. Thanks again Amos.
